File: client/csharp/src/Stream.cs
using System;
using System.Threading;
using KRPC.Schema.KRPC;

namespace KRPC.Client
{
    /// <summary>
    /// Object representing a stream.
    /// </summary>
    public class Stream<TReturnType> : IEquatable<Stream<TReturnType>>
    {
        internal readonly StreamImpl stream;

        internal Stream (Connection connection, ulong id)
        {
            stream = connection.StreamManager.GetStream (typeof(TReturnType), id);
        }

        internal Stream (Connection connection, ProcedureCall call)
        {
            stream = connection.StreamManager.AddStream (typeof(TReturnType), call);
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public override bool Equals (object obj)
        {
            if (ReferenceEquals (this, obj))
                return true;
            if (ReferenceEquals (obj, null))
                return false;
            var typedObj = obj as Stream<TReturnType>;
            return typedObj != null && Equals (typedObj);
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public bool Equals (Stream<TReturnType> other)
        {
            return !ReferenceEquals (other, null) && stream.Id == other.stream.Id;
        }

        /// <summary>
        /// Returns true if the objects are equal.
        /// </summary>
        public static bool operator == (Stream<TReturnType> lhs, Stream<TReturnType> rhs)
        {
            if (ReferenceEquals (lhs, null) || ReferenceEquals (rhs, null))
                return ReferenceEquals (lhs, rhs);
            if (ReferenceEquals (lhs, rhs))
                return true;
            return lhs.Equals (rhs);
        }

        /// <summary>
        /// Returns true if the objects are not equal.
        /// </summary>
        public static bool operator != (Stream<TReturnType> lhs, Stream<TReturnType> rhs)
        {
            if (ReferenceEquals (lhs, null) || ReferenceEquals (rhs, null))
                return !ReferenceEquals (lhs, rhs);
            if (ReferenceEquals (lhs, rhs))
                return false;
            return !(lhs.Equals (rhs));
        }

        /// <summary>
        /// Hash code for the object.
        /// </summary>
        public override int GetHashCode ()
        {
            return stream.GetHashCode ();
        }

        /// <summary>
        /// Start the stream.
        /// </summary>
        public void Start(bool wait = true) {
            if (stream.Started)
                return;
            if (!wait) {
                stream.Start();
            } else {
                var condition = stream.Condition;
                lock (condition) {
                    stream.Start();
                    Monitor.Wait (condition);
                }
            }
        }

        /// <summary>
        /// The rate of the stream, in Hertz.
        /// </summary>
        public float Rate {
            get { return stream.Rate; }
            set { stream.Rate = value; }
        }

        /// <summary>
        /// The most recent value of the stream.
        /// </summary>
        public TReturnType Get () {
            if (!stream.Started)
                Start();
            var value = stream.Value;
            var exn = value as System.Exception;
            if (exn != null)
                throw exn;
            return (TReturnType) value;
        }

        /// <summary>
        /// Condition variable that is notified when the streams value changes.
        /// </summary>
        public object Condition {
            get { return stream.Condition; }
        }

        /// <summary>
        /// Wait until the next stream update.
        /// </summary>
        public void Wait (double timeout = -1) {
            if (!stream.Started)
                stream.Start();
            if (timeout >= 0)
                Monitor.Wait (stream.Condition, (int)(timeout*1000.0));
            else
                Monitor.Wait (stream.Condition);
        }

        /// <summary>
        /// Add a callback that is invoked whenever the stream is updated.
        /// </summary>
        public int AddCallback (Action<TReturnType> callback) {
            return stream.AddCallback ((object x) => callback((TReturnType)x));
        }

        /// <summary>
        /// Remove a callback, based on its tag.
        /// </summary>
        public void RemoveCallback (int tag) {
          stream.RemoveCallback (tag);
        }

        /// <summary>
        /// Remove the stream from the server.
        /// </summary>
        public void Remove ()
        {
            stream.Remove ();
        }
    }
}

File: client/python/krpc/stream.py
from __future__ import annotations
import threading
from typing import Any, Callable, Optional, TYPE_CHECKING
from krpc.types import TypeBase
import krpc.schema.KRPC_pb2 as KRPC
if TYPE_CHECKING:
    from krpc.client import Client
    from krpc.streammanager import StreamImpl


class Stream:
    """ A streamed remote procedure call. When called, returns the
        most recently received result of the call. """

    def __init__(self, stream: StreamImpl) -> None:
        self._stream = stream

    @classmethod
    def from_stream_id(cls, client: Client, stream_id: int,
                       return_type: TypeBase) -> Stream:
        """ Create a stream from an existing stream id on the server """
        stream = client._stream_manager.get_stream(return_type, stream_id)
        return cls(stream)

    @classmethod
    def from_call(cls, client: Client, return_type: TypeBase, call: KRPC.ProcedureCall) -> Stream:
        """ Create a stream from a remote procedure call """
        stream = client._stream_manager.add_stream(return_type, call)
        return cls(stream)

    def start(self, wait: bool = True) -> None:
        """ Start the stream. If wait is true,
            blocks until the stream has received its first update. """
        if self._stream.started:
            return
        if not wait:
            self._stream.start()
        else:
            with self._stream.condition:
                self._stream.start()
                self._stream.condition.wait()

    @property
    def rate(self) -> float:
        """ The update rate for the stream in Hertz.
            Zero if the rate is unlimited. """
        return self._stream.rate

    @rate.setter
    def rate(self, value: float) -> None:
        """ The update rate for the stream in Hertz.
            Zero if the rate is unlimited. """
        self._stream.rate = value

    def __call__(self):  # type: ignore[no-untyped-def]
        """ Get the most recent value for this stream. """
        if not self._stream.started:
            self.start()
        value = self._stream.value
        if isinstance(value, Exception):
            raise value
        return value

    @property
    def condition(self) -> threading.Condition:
        """ Condition variable that is notified when the stream updates. """
        return self._stream.condition

    def wait(self, timeout: Optional[float] = None) -> None:
        """ Wait until the next stream update or a timeout occurs.
            The condition variable must be locked before calling this method.

            When timeout is not None, it should be a floating point number
            specifying the timeout in seconds for the operation. """
        if not self._stream.started:
            self._stream.start()
        self._stream.condition.wait(timeout=timeout)

    def add_callback(self, callback: Callable[[Any], None]) -> None:
        """ Add a callback that is invoked whenever the stream is updated. """
        self._stream.add_callback(callback)

    def remove_callback(self, callback: Callable[[Any], None]) -> None:
        """ Remove a callback. """
        self._stream.remove_callback(callback)

    def remove(self) -> None:
        """ Remove the stream """
        self._stream.remove()
